File: src/aside/java/org/taktik/icure/dao/HealthcarePartyDAO.java
/*
 * Copyright (C) 2018 Taktik SA
 *
 * This file is part of iCureBackend.
 *
 * iCureBackend is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as published by
 * the Free Software Foundation.
 *
 * iCureBackend is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with iCureBackend.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.taktik.icure.dao;


import org.taktik.icure.db.PaginatedList;
import org.taktik.icure.entities.HealthcareParty;
import org.taktik.icure.db.PaginationOffset;

import java.util.List;
import java.util.Map;

public interface HealthcarePartyDAO extends GenericDAO<HealthcareParty>{

	PaginatedList<HealthcareParty> findByHcPartyNameContainsFuzzy(String searchString, PaginationOffset offset, Boolean desc);

	List<HealthcareParty> findHealthcareParties(String searchString, int offset, int limit);

    /**
     * Returns a map indexed by owners of all HcPartyKeys encrypted using healthcarePartyId
     * private key.
     *
     * @param healthcarePartyId
     * @return the map of keys indexed by owner
     */
	Map<String,String> getHcPartyKeysForDelegate(String healthcarePartyId);

	List<HealthcareParty> findByNihii(String nihii);

	List<HealthcareParty> findBySsin(String ssin);

	PaginatedList<HealthcareParty> findBySpecialityPostcode(String type, String spec, String startCode, String endCode);

	PaginatedList<HealthcareParty> listHealthCareParties(PaginationOffset pagination, Boolean desc);

	List<HealthcareParty> findByName(String name);

	PaginatedList<HealthcareParty> findBySsinOrNihii(String searchValue, PaginationOffset<String> paginationOffset, Boolean desc);

	List<HealthcareParty> findByParentId(String parentId);
}

File: src/main/kotlin/org/taktik/icure/asyncdao/HealthcarePartyDAO.kt
/*
 *  iCure Data Stack. Copyright (c) 2020 Taktik SA
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as
 *     published by the Free Software Foundation, either version 3 of the
 *     License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful, but
 *     WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *     Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public
 *     License along with this program.  If not, see
 *     <https://www.gnu.org/licenses/>.
 */

package org.taktik.icure.asyncdao

import kotlinx.coroutines.flow.Flow
import org.taktik.couchdb.ViewQueryResultEvent
import org.taktik.icure.db.PaginationOffset
import org.taktik.icure.entities.HealthcareParty
import org.taktik.icure.entities.embed.Identifier

interface HealthcarePartyDAO : GenericDAO<HealthcareParty> {
	fun listHealthcarePartiesByNihii(nihii: String?): Flow<HealthcareParty>

	fun listHealthcarePartiesBySsin(ssin: String): Flow<HealthcareParty>

	fun listHealthcarePartiesBySpecialityAndPostcode(type: String, spec: String, firstCode: String, lastCode: String): Flow<ViewQueryResultEvent>

	fun findHealthCareParties(pagination: PaginationOffset<String>, desc: Boolean?): Flow<ViewQueryResultEvent>

	fun listHealthcarePartiesByName(name: String): Flow<HealthcareParty>

	fun findHealthcarePartiesBySsinOrNihii(searchValue: String?, offset: PaginationOffset<String>, desc: Boolean?): Flow<ViewQueryResultEvent>

	fun findHealthcarePartiesByHcPartyNameContainsFuzzy(searchString: String?, offset: PaginationOffset<String>, desc: Boolean?): Flow<ViewQueryResultEvent>

	fun listHealthcareParties(searchString: String, offset: Int, limit: Int): Flow<HealthcareParty>

	@Deprecated(message = "A HCP may now have multiple AES Keys. Use getAesExchangeKeysForDelegate instead")
	suspend fun getHcPartyKeysForDelegate(healthcarePartyId: String): Map<String, String>

	suspend fun getAesExchangeKeysForDelegate(healthcarePartyId: String): Map<String, List<String>>

	fun listHealthcarePartiesByParentId(parentId: String): Flow<HealthcareParty>

	fun findHealthcarePartiesByIds(hcpIds: Flow<String>): Flow<ViewQueryResultEvent>
	fun listHealthcarePartyIdsByIdentifiers(hcpIdentifiers: List<Identifier>): Flow<String>
	fun listHealthcarePartyIdsByCode(codeType: String?, codeCode: String?): Flow<String>
	fun listHealthcarePartyIdsByTag(tagType: String?, tagCode: String?): Flow<String>
}
